src/Microsoft.Diagnostics.DebugServices.Implementation/Host.cs

```diff
@@ -0,0 +1,89 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using Microsoft.Diagnostics.DebugServices;
+
+namespace Microsoft.Diagnostics.DebugServices.Implementation
+{
+    public class Host : IHost
+    {
+        private readonly ServiceManager _serviceManager;
+        private ServiceContainer _serviceContainer;
+        private readonly List<ITarget> _targets = new();
+        private int _targetIdFactory;
+
+        public Host(HostType type)
+        {
+            HostType = type;
+
+            _serviceManager = new ServiceManager();
+
+            // Register all the services and commands in the Microsoft.Diagnostics.DebugServices.Implementation assembly
+            _serviceManager.RegisterAssembly(typeof(Target).Assembly);
+        }
+
+        /// <summary>
+        /// Service manager
+        /// </summary>
+        public ServiceManager ServiceManager => _serviceManager;
+
+        /// <summary>
+        /// Service container
+        /// </summary>
+        public ServiceContainer ServiceContainer => _serviceContainer ?? throw new InvalidOperationException();
```

Comment:
The ServiceContainer is only created after all the extensions are loaded and all the service factories are added to the ServiceManager. So this null check is a safety check if some code tries to get a service before things are ready.  We want to be explicit about when the ServiceContainer is ready because after the it is created we call no longer load any extensions which add service factories.

Reply:
I didn't answer the first question: IHost.Services is the public abstracted IServiceProvider API.  ServiceContainer (even though it is public) exposes more properties/methods to the host implementations (dotnet-dump analyze, SOS.Extensions, TestHost, etc).